src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py

```diff
@@ -739,66 +733,149 @@ def _visit_map(self, node: gtir.FunCall) -> ValueExpr:
-        # We store the value of reduce identity in the visitor context to have it
-        # available while visiting the input to reduction; this value might be used
-        # by the `neighbors` visitor to fill the skip values in the neighbors list.
-        prev_reduce_identity = self.reduce_identity
-        self.reduce_identity = reduce_identity
-
-        try:
-            input_expr = self.visit(node.args[0])
-        finally:
-            # ensure that we leave the visitor in the same state as we entered
-            self.reduce_identity = prev_reduce_identity
-
-        assert isinstance(input_expr, MemletExpr | ValueExpr)
-        input_desc = input_expr.dc_node.desc(self.sdfg)
-        assert isinstance(input_desc, dace.data.Array)
-
-        if len(input_desc.shape) > 1:
-            assert isinstance(input_expr, MemletExpr)
-            ndims = len(input_desc.shape) - 1
-            # the axis to be reduced is always the last one, because `reduce` is supposed
-            # to operate on `ListType`
-            assert set(input_expr.subset.size()[0:ndims]) == {1}
-            reduce_axes = [ndims]
+        assert input_expr.local_offset is not None
+        connectivity = dace_utils.connectivity_identifier(input_expr.local_offset)
+        connectivity_node = self.state.add_access(connectivity)
+        connectivity_desc = connectivity_node.desc(self.sdfg)
+        connectivity_desc.transient = False
+
+        desc = input_expr.dc_node.desc(self.sdfg)
+        if isinstance(input_expr, MemletExpr):
+            local_dim_indices = [i for i, size in enumerate(input_expr.subset.size()) if size != 1]
         else:
-            reduce_axes = None
+            local_dim_indices = list(range(len(desc.shape)))
+
+        assert len(local_dim_indices) == 1
+        local_dim_index = local_dim_indices[0]
+        assert desc.shape[local_dim_index] == offset_provider.max_neighbors
+
+        # we lower the reduction map with WCR out memlet in a nested SDFG
+        nsdfg = dace.SDFG("reduce_with_skip_values")
+        nsdfg.add_array(
+            "values",
+            (desc.shape[local_dim_index],),
+            desc.dtype,
+            strides=(desc.strides[local_dim_index],),
+        )
+        nsdfg.add_array(
+            "neighbor_indices",
+            (connectivity_desc.shape[1],),
+            connectivity_desc.dtype,
+            strides=(connectivity_desc.strides[1],),
+        )
+        nsdfg.add_scalar("acc", desc.dtype)
+        st_init = nsdfg.add_state("init")
+        st_init.add_edge(
+            st_init.add_tasklet(
+                "init_acc",
+                {},
+                {"__val"},
+                f"__val = {reduce_init.dc_dtype}({reduce_init.value})",
+            ),
+            "__val",
+            st_init.add_access("acc"),
+            None,
+            dace.Memlet(data="acc", subset="0"),
+        )
+        st_reduce = nsdfg.add_state_after(st_init, "reduce")
+        st_reduce.add_mapped_tasklet(
```

Comment:
Your map has default schedule, which depending on the situation might be turned into a parallel schedule, which requires atomic reduction.
You should explicitly set it to sequential, which is a reasonable assumption for GT4Py (see later).
In that case you should also set `wcr_nonatomic` (this is a Memlet property) to `True`.
If you do that please add the following todo as well
```
TODO(phimuell): decide if auto optimizer should reset `wcr_nonatomic` properties, as DaCe does it.
```

Note if you follow my suggestions above, reductions along any dimension will be sequential, which is according to my understanding of Gt4Py the majority of cases, as you reduce along a local dimension.
However, this also applies to reduction along every dimensions not only locals but horizontal or vertical dimensions too!
This is certainly okay for local and probably to some degree and in some special situations for vertical, but certainly not for horizontal dimensions.
For that reason I would add a switch that checks (for now) if the reduction is along a local dimension and then do what you are doing now or otherwise puts a reduction library node there.

Furthermore, because you collect them in a local array, the memory loading is not combined, i.e. the address thread 2 loads is not adjacent to thread 1.
For that you would need to collect the data in a fundamentally different way, but for now this should be fine. 
However, this has no priority (please add some todo if you agree with me) and we should talk with Felix/Hannes/Yannis before we do that.
In the long run we will have to add this into a library node, or at least make a custom expansion.

Reply:
I don't think we need to set the map schedule and the `wcr_nonatomic`. It is up to DaCe to decide the schedule for this loop and, based on that decision, whether the WCR should use atomics or not.


Note that we will never encounter skip values on global dimensions. Currently, the lowering supports only reduction along a single dimension (corresponding to the neighbors local dimension) and there is already a check for that: `assert len(local_dim_indices) == 1`
I could turn the assert into an exception.

Reply:
This are exactly the points that are hard to find out on the SDFG level and a lot can go wrong.
Why spend compute time to figuring out something that can be set correctly from construction.
Here GT4Py has to help the optimization, because recovering the information is super hard and it would be much simpler, just to emit the right/smart constructs.
On the SDFG there is no fool prove way to figuring out what a local dimension is, on which you can make certain assumptions.
This is the whole point of having a DSL to have more information what you want to do, we have to put that knowledge to work.

Reply:
I think DaCe just looks at the hierarchy of the maps and applies the parallel schedule (GPU global or CPU OpenMP) on the outerst map. If the schedule is not set, the map will be sequential. This is case of all inner maps with the exception of GPU tiling transformation, probably. The setting of the map schedule is done at the very end, I think in code generation for CPU (but I would have to check). So this kind of reduce maps, that operate on a local dimension, will by definition always be inner maps and therefore be left with sequential schedule.

Anyway, I also get your point and I don't have a strong opinion against setting the schedule to sequential, because anyway that is what we want to achieve in the end.

Reply:
Thanks.


This morning I realized something but I am not sure if it is actually done.
But you do reduction in one swoop, i.e. the replacement of skip values and the reduction is done in the same Tasklet?
It is not that first an array is build filled and then reduced.
If so the order must be changed, because then the initialization between the two loops blocks their fusing.

I am still thinking that this should be a library node, at least in the long run.
Because then we could at expansion time, put the initialization at the beginning of the scope, and avoid putting the reduction loop inside a nested SDFG.
This would allow to fuse it, because so the node can not be fused.

Reply:
I totally agree about the usage of a library node, in the long-term, not only for reduce but also for neighbors and map. In this way we could transform and eventually fuse different expressions on local dimensions.

About the current implementation. The tasklet just writes a scalar value, either the input value or the reduce identity, into the accumulator scalar node. This write memlet uses a WCR to accumulate the result, rather than overwriting it.

> It is not that first an array is build filled and then reduced.

Unfortunately not. This was the original implementation, before I found a bug in the MapReduceFusion transformation which is used in the auto-optimize pipeline (you should maybe consider to remove it, it contains a bug). The bug is discussed here: https://chat.spcl.inf.ethz.ch/spcl/pl/8mtgtqjb378hfy7h9a96sy3nhc

> If so the order must be changed, because then the initialization between the two loops blocks their fusing.

Which two loops?

Reply:
> you should maybe consider to remove it

I actually wanted to do that. I will make a PR right away.

> Which two loops?

This was a reference to the original implementation.
Where we had a collecting loop and a reduction loop.
Because, the nested SDFG was in between the two could not be fused together.